command-derive/command-derive-impl/src/lib.rs

```diff
@@ -71,92 +61,204 @@ pub fn command_derive(input: TokenStream) -> TokenStream {
+                            ),
+                            true => (
+                                quote!{ res.extend(#fld_ident.into_iter()) },
+                                quote!{ ::command_derive::FromStrIter::from_str_iter(&mut args)? }
+                            )
+                        }
+                    })
+                    .unzip();
+                let idx_fields_piece: Vec<_> =
+                    (0..field_num).map(|i| Member::Unnamed(i.into())).collect();
+                let pattern = quote! {
+                    #ident (#(#named_fields,)*)
+                };
+                (
+                    named_fields_to_str_piece,
+                    read_fields_piece,
+                    idx_fields_piece,
+                    pattern,
+                )
+            }
+            Fields::Unit => (Vec::new(), Vec::new(), Vec::new(), quote! { #ident }),
         };
         var_idents.push(ident);
         codes.push(code);
-        patterns.push(pattern);
-        named_fields.push(named_fields_piece);
+        named_fields_to_str.push(named_fields_to_str_piece);
+        read_fields.push(read_fields_piece);
         idx_fields.push(idx_fields_piece);
+        patterns.push(pattern);
     }
-    (quote!{
-impl ::command_derive::Command for #enum_ident {
-    fn ident(&self) -> &'static str {
-        match self {
-            #(
-                #enum_ident::#var_idents { .. } => #codes,
-            )*
-                _ => panic!("`ident()` on skipped variant")
+    (quote! {
+    impl ::command_derive::Command for #enum_ident {
+        fn ident(&self) -> &'static str {
+            match self {
+                #(
+                    #enum_ident::#var_idents { .. } => #codes,
+                )*
+            }
+        }
+
+        fn extract_args(&self) -> Vec<String> {
+            use ::core::iter::IntoIterator;
+
+            let mut res = Vec::new();
+            match self {
+                #(
+                    #enum_ident::#patterns => {
+                        #(#named_fields_to_str;)*
+                    },
+                )*
+            }
+            res
+        }
+
+        fn from_protocol<I, S>(code: &str, mut args: I) -> Result<Self, ::anyhow::Error>
+        where
+            I: Iterator<Item = S>,
+            S: AsRef<str>
+        {
+            let res = match code {
+                #(
+                    #codes => #enum_ident::#var_idents{#(
+                        #idx_fields: #read_fields,
+                    )*},
+                )*
+                code => return Err(::anyhow::anyhow!("Unknown command code: {}", code))
+            };
+            if args.next().is_some() {
+                return Err(::anyhow::anyhow!("Too many args"));
+            }
+            Ok(res)
         }
     }
+        })
+    .into()
+}
+
+// Derives `IntoIterator<IntoIter=Vec<String>::IntoIter>` for `&Self`
+// and `FromStrIter` for `Self`
+// Applyable for struct-s
```

Comment:
Should be a `///` doc comment I suppose 
```suggestion
/// Derives `IntoIterator<IntoIter=Vec<String>::IntoIter>` for `&Self`
/// and `FromStrIter` for `Self`
/// Applyable for struct-s
```